server/src/main/java/org/opensearch/common/util/PlainIterator.java

```diff
@@ -44,29 +45,26 @@
 public class PlainIterator<T> implements Iterable<T>, Countable {
     private final List<T> elements;
 
-    // Calls to nextOrNull might be performed on different threads in the transport actions so we need the volatile
-    // keyword in order to ensure visibility. Note that it is fine to use `volatile` for a counter in that case given
-    // that although nextOrNull might be called from different threads, it can never happen concurrently.
-    private volatile int index;
+    private AtomicInteger index = new AtomicInteger();
```

Comment:
The comment here explains why the original author thought a volatile primitive is acceptable. Why do you think that is incorrect?

Also, the primitive int will use 4 bytes, while an AtomicInteger object will use at least 16 bytes (if my understanding of how Java allocates Objects is correct...), so this change will result in at least a small increase in memory footprint. The difference is may be completely negligible but I don't know that for sure.

Reply:
> Also, the primitive int will use 4 bytes, while an AtomicInteger object will use at least 16 bytes (if my understanding of how Java allocates Objects is correct...)

It's worse than that. The 16 bytes is just for the object, but it needs to maintain the internal 4-byte value, plus objects are 8-byte aligned so it'll be 24 bytes.  [Some guy spent way too long investigating this a decade ago](https://stackoverflow.com/a/8959355/1161484).

Reply:
@andrross 
"volatile" keyword means you want to see the actual value in multithreaded env - i.e. no caching of the value.
But operations like num++ and ++num aren't atomic. And the other thread can see the outdated value.

Reply:
@dk2k As you say, atomicity is required for incrementing, but that never happens in a multithreaded context (per the comment).

All that is required for reading is getting the latest value.  If you catch it in the middle of an update you may get +1 or -1 but that's really no different than grabbing it right before or right after an (atomic) update as far as the other thread is concerned.  

If you think the original author's claim that `nextOrNull` won't ever be called concurrently is wrong, please explain why.